homeassistant/components/sia/sia_entity_base.py

```diff
@@ -0,0 +1,131 @@
+"""Module for SIA Base Entity."""
+from __future__ import annotations
+
+from abc import abstractmethod
+import logging
+from typing import Any
+
+from pysiaalarm import SIAEvent
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PORT
+from homeassistant.core import CALLBACK_TYPE, State, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.event import async_call_later
+from homeassistant.helpers.restore_state import RestoreEntity
+
+from .const import CONF_ACCOUNT, CONF_PING_INTERVAL, DOMAIN, SIA_EVENT, SIA_NAME_FORMAT
+from .utils import get_attr_from_sia_event, get_unavailability_interval
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SIABaseEntity(RestoreEntity):
+    """Base class for SIA entities."""
+
+    def __init__(
+        self,
+        entry: ConfigEntry,
+        account_data: dict[str, Any],
+        zone: int,
+        device_class: str,
+    ) -> None:
+        """Create SIABaseEntity object."""
+        self._entry: ConfigEntry = entry
+        self._account_data: dict[str, Any] = account_data
+        self._zone: int = zone
+        self._attr_device_class: str = device_class
+
+        self._port: int = self._entry.data[CONF_PORT]
+        self._account: str = self._account_data[CONF_ACCOUNT]
+        self._ping_interval: int = self._account_data[CONF_PING_INTERVAL]
+
+        self._cancel_availability_cb: CALLBACK_TYPE | None = None
+
+        self._attr_extra_state_attributes: dict[str, Any] = {}
+        self._attr_should_poll = False
+        self._attr_name = SIA_NAME_FORMAT.format(
+            self._port, self._account, self._zone, self._attr_device_class
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass.
+
+        Overridden from Entity.
+
+        1. register the dispatcher and add the callback to on_remove
+        2. get previous state from storage and pass to entity specific function
+        3. if available: create availability cb
+        """
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                SIA_EVENT.format(self._port, self._account),
+                self.async_handle_event,
+            )
+        )
+        self.handle_last_state(await self.async_get_last_state())
+        if self._attr_available:
+            self.async_create_availability_cb()
+
+    @abstractmethod
+    def handle_last_state(self, last_state: State | None) -> None:
+        """Handle the last state."""
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Run when entity will be removed from hass.
+
+        Overridden from Entity.
+        """
+        if self._cancel_availability_cb:
+            self._cancel_availability_cb()
+
+    async def async_handle_event(self, sia_event: SIAEvent) -> None:
```

Comment:
This doesn't need to be a coroutine function since we don't await inside. We can make it a callback.